// m.config/lib/config.js

module.exports = {

  hmbaidu: {
    "activityYdlServer": "4446f7fa4186990430cb61ee85ce0d00",
    "activityYidianlingServer": "4446f7fa4186990430cb61ee85ce0d00",
    "activityHaoshi99Server": "4446f7fa4186990430cb61ee85ce0d00",
    "activityYdlcdnServer": "4446f7fa4186990430cb61ee85ce0d00",
    "dev.activity.yidianling.com:8080": "4446f7fa4186990430cb61ee85ce0d00",
  },

  servers: {
    "0": "http://47.97.49.44:8888/api",
    "1": "http://47.97.49.44:8888/api",
    "2": "http://47.97.49.44:8888/api",
    "7": "http://47.97.49.44:8888/api",
    "9": "https://api.ydl.com/api",
  },

  jump: {
    "0": "http://testm.yidianling.com",
    "1": "http://testm.yidianling.com",
    "2": "http://testm.yidianling.com",
    "7": "http://testm.yidianling.com",
    "9": "https://m.yidianling.com",
  },

  wechat: {
    appId: "wx2723a331ca027fd9",
    secret: "e54b0a6c957ebe678395aec05d82dbb1",
  },

  qiniu: {
    "accessKey": "1tG4V12sMejSc0U2-rtMeny8tSlAc1f4K0VZy9rM",
    "secretKey": "wLN9AgUinMZ8Alan3xyrQzw_jZyMcHBDVhz7hL7Y",
    "bucket": "imgqiniu-yidianling-com",
  },

	G: {
		env: Config.env,
    time: Config.time,
    version: Config.version,
    dist: Config.dist,
    webServer: Config.webServer,
    nodeServer: Config.nodeServer,
    jslist: {
      "zoosnet": {
        "url": "https://dat.zoosnet.net/js/lsjs.aspx?siteid=DAT83236505&float=1&lng=cn",
      }
    },
	},

	headers(res){
		res.set('Content-type', 'application/json');
		res.set('Access-Control-Allow-Origin', '*');
		res.set("Access-Control-Allow-Headers", 'X-Requested-With');
	},

	requestHeader: {
		headers: {
			"User-Agent": "ydl111",
			"source": "ydl",
			"appId": "ydl",
			"version": "1.0",
			"authentication": ""
		}
	},

	timeout(res){

    res.setTimeout(30000, () => {
      console.log("响应超时");
      res.json(Object.assign({}, this.dataJson, {"code": "-1", "msg": "响应超时"}));
      return;
    });
	},

	dataJson: {
		code: 0,
		msg: "success",
		data: [],
	}
}
